File: tools-manage/invocationlab-admin/src/main/java/org/javamaster/invocationlab/admin/service/registry/impl/AbstractRegisterFactory.java
package org.javamaster.invocationlab.admin.service.registry.impl;

import org.javamaster.invocationlab.admin.service.registry.Register;
import org.javamaster.invocationlab.admin.service.registry.RegisterFactory;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * @author yudong
 */
public abstract class AbstractRegisterFactory implements RegisterFactory {
    protected Map<String, Register> allRegisters = new HashMap<>();
    protected Set<String> clusterSet = new HashSet<>();

    @Override
    public void addCluster(String cluster) {
        clusterSet.add(cluster);
    }

    @Override
    public Register get(String cluster) {
        if (allRegisters.containsKey(cluster)) {
            return allRegisters.get(cluster);
        }
        Register register = create(cluster);
        allRegisters.put(cluster, register);
        return register;
    }

    @Override
    public Register remove(String cluster) {
        return allRegisters.remove(cluster);
    }

    public abstract Register create(String cluster);

    @Override
    public Set<String> getClusterSet() {
        return clusterSet;
    }
}

File: tools-manage/invocationlab-admin-21/src/main/kotlin/org/javamaster/invocationlab/admin/service/registry/impl/AbstractRegisterFactory.kt
package org.javamaster.invocationlab.admin.service.registry.impl

import org.javamaster.invocationlab.admin.service.registry.Register
import org.javamaster.invocationlab.admin.service.registry.RegisterFactory

/**
 * @author yudong
 */
abstract class AbstractRegisterFactory : RegisterFactory {
    private var allRegisters: MutableMap<String, Register> = HashMap()
    private var clusterSet: MutableSet<String> = HashSet()

    override fun addCluster(cluster: String) {
        clusterSet.add(cluster)
    }

    override fun get(cluster: String): Register {
        if (allRegisters.containsKey(cluster)) {
            return allRegisters[cluster]!!
        }
        val register = create(cluster)
        allRegisters[cluster] = register
        return register
    }

    override fun remove(cluster: String): Register {
        return allRegisters.remove(cluster)!!
    }

    abstract fun create(cluster: String): Register

    override fun getClusterSet(): MutableSet<String> {
        return clusterSet
    }
}
